statsmodels/regression/_tools.py

```diff
@@ -0,0 +1,118 @@
+    Minimal implementation of WLS optimized for performance.
+
+    Parameters
+    ----------
+    endog : array-like
+        1-d endogenous response variable. The dependent variable.
+    exog : array-like
+        A nobs x k array where `nobs` is the number of observations and `k`
+        is the number of regressors. An intercept is not included by default
+        and should be added by the user. See
+        :func:`statsmodels.tools.add_constant`.
+    weights : array-like, optional
+        1d array of weights.  If you supply 1/W then the variables are pre-
+        multiplied by 1/sqrt(W).  If no weights are supplied the default value
+        is 1 and WLS reults are the same as OLS.
+
+    Notes
+    -----
+    Need resid, scale, fittedvalues, model.weights!
+        history['scale'].append(tmp_results.scale)
+        if conv == 'dev':
+            history['deviance'].append(self.deviance(tmp_results))
+        elif conv == 'sresid':
+            history['sresid'].append(tmp_results.resid/tmp_results.scale)
+        elif conv == 'weights':
+            history['weights'].append(tmp_results.model.weights)
+    Does not perform and checks on the input data
+    """
+
+    def __init__(self, endog, exog, weights=1.0):
+        self.endog = endog
+        self.exog = exog
+        self.weights = weights
+        w_half = np.sqrt(weights)
+
+        self.wendog = w_half * endog
+        if np.isscalar(weights):
+            self.wexog = w_half * exog
+        else:
+            self.wexog = w_half[:, None] * exog
+
+    def fit(self, cov=True, method='pinv'):
+        """
+        Minimal implementation of WLS optimized for performance.
+
+        Parameters
+        ----------
+        cov : bool, optional
+            Flag indicating whether to the compute normalized covariance
+        method : str, optional
+            Method to use to estimate parameters.  "pinv", "qr" or "lstsqr"
+
+              * "pinv" uses the Moore-Penrose pseudoinverse
+                 to solve the least squares problem.
+              * "qr" uses the QR factorization.
+              * "lstsq" uses the least squares implementation in SciPy
+
+        Returns
+        -------
+        results : namedtuple
+            Named tuple containing the fewest terms needed to implement
+            iterative estimation in models. Currently
+
+              * params : Estimated parameters
+              * fittedvalues : Fit values using original data
+              * resid : Residuals using original data
+              * normalized_cov_params
+              * model : namedtuple with one field, weights
+              * scale : scale computed using weighted residuals
+
+        Notes
+        -----
+        Does not perform and checks on the input data
+
+        See Also
+        --------
+        statsmodels.regression.linear_model.WLS
+        """
+        normalized_cov_params = None
+        if method == 'pinv':
+            pinv_wexog = np.linalg.pinv(self.wexog)
```

Comment:
Is there a way that we could `pinv` on `self.exog` only once? Seems like this will run ever iteration and the result shoudl be same every time, if I understand correctly.